File: FigureScripts/ODE_SEnIV.m
% Code by Marian Dominguez-Mirazo and David Demory, 2023 
% See Methods for the mathematical description of the ODE model


function dy = ODE_SEnIV(t,y,params)

dy = zeros(params.n+3,1);

% Special case when there are no E compartments
if params.n==0
    
    % n = 0
    S = y(1);         % Susceptible cell
    I = y(2);         % Infected
    V = y(3);         % Free virus
    
    %% ODE system
    dy(1) = params.mu*S.*(1-(S+I)./params.K) - params.phi*S.*V;
    dy(2) = params.phi*S.*V - params.eta*I;
    dy(3) = params.beta*params.eta*I - params.phi*S*V;
    
else
    
    % n > 0
    S = y(1);           % Susceptible cell
    E = y(2:params.n+1);       % n Exposed cell classes
    I = y(params.n+2);         % Infected
    V = y(params.n+3);         % Free virus
    
    % funtions transfer
    rate_in = eye(params.n).*repmat((params.n+1)*params.eta,params.n,1);
    rate_in(1,1) = params.phi*S.*V;
    
    E_in  = [1;E(1:params.n-1)];
    
    %% ODE system
    %
    % Susceptible cells
    dy(1) = params.mu*S.*(1-(S+I+sum(E))./params.K) - params.phi*S.*V;
    dy(2:params.n+1) = rate_in*E_in  - (params.n+1)*params.eta*E;
    dy(params.n+2) = (params.n+1)*params.eta*E(end) - (params.n+1)*params.eta*I;
    dy(params.n+3) = params.beta*(params.n+1)*params.eta*I - params.phi*S*V;
end

end

File: ParameterInference/Functions/ODE_SEnIV.jl
# Code by Marian Dominguez-Mirazo and David Demory, 2023
# See Methods for the mathematical description of the ODE model

function ODE_SEnIV(du,u,p,t)
    mu, K, phi, beta, eta, n = p[1:6]
    n = Int64(n)
    
    # Special case when there are no E compartments
    if n ==0
        S,I,V = u
        du[1] = mu*S*(1-(S+I)/K) - phi*S*V
        du[2] = phi*S*V - eta*I
        du[3] = beta*eta*I - phi*S*V
        
    else
        rate_in = zeros(eltype(phi),n,n)
        S = u[1]
        E = u[2:(n+1)]
        I = u[n+2]
        V = u[n+3]
        
        for i = 1:n
            rate_in[i,i] = (n+1)*eta
        end
        rate_in[1,1] = phi*S*V
        E_in  = [1;E[1:(n-1)]]
        
        du[1] = mu*S*(1-(S+I+sum(E))/K) - phi*S*V
        du[2:(n+1)] = rate_in*E_in - (n+1)*eta*E
        du[n+2] = (n+1)*eta*E[end] - (n+1)*eta*I
        du[n+3] = beta*(n+1)*eta*I - phi*S*V
        
    end
    
    return du
end